geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+    xxs: NDArrayNum, num_chunks: tuple[int, int], nb_valids_per_block: list[int]
+) -> list[list[int]]:
+    """
+    Get list of 1D valid subsample indices relative to the block for each block.
+
+    The 1D valid subsample indices correspond to the subsample index to apply for a flattened array of valid values.
+    Relative to the block means converted so that the block indexes for valid values starts at 0 up to the number of
+    valid values in that block (while the input indices go from zero to the total number of valid values in the full
+    array).
+
+    :param xxs: Subsample 1D indexes among a total number of valid values.
+    :param num_chunks: Number of chunks in X and Y.
+    :param nb_valids_per_block: Number of valid pixels per block.
+
+    :returns: Relative 1D valid subsample index per block.
+    """
+
+    # Apply a cumulative sum to get the first 1D total index of each block
+    valids_cumsum = np.cumsum(nb_valids_per_block)
+
+    # We can write a faster algorithm by sorting
+    xxs = np.sort(xxs)
+
+    # TODO: Write nested lists into array format to further save RAM?
+    # We define a list of indices per block
+    relative_ind_per_block = [[] for _ in range(num_chunks[0] * num_chunks[1])]
+    k = 0  # K is the block number
+    for x in xxs:
+
+        # Move to the next block K where current 1D subsample index is, if not in this one
+        while x >= valids_cumsum[k]:
+            k += 1
+
+        # Add 1D subsample index  relative to first subsample index of this block
+        first_xindex_block = valids_cumsum[k - 1] if k >= 1 else 0  # The first 1D valid subsample index of the block
+        relative_xindex = x - first_xindex_block
+        relative_ind_per_block[k].append(relative_xindex)
+
+    return relative_ind_per_block
+
+
+@dask.delayed  # type: ignore
+def _delayed_nb_valids(arr_chunk: NDArrayNum) -> NDArrayNum:
+    """Count number of valid values per block."""
+    return np.array([np.count_nonzero(np.isfinite(arr_chunk))]).reshape((1, 1))
+
+
+@dask.delayed  # type: ignore
+def _delayed_subsample_block(arr_chunk: NDArrayNum, subsample_indices: NDArrayNum) -> NDArrayNum:
+    """Subsample the valid values at the corresponding 1D valid indices per block."""
+
+    s_chunk = arr_chunk[np.isfinite(arr_chunk)][subsample_indices]
+
+    return s_chunk
+
+
+@dask.delayed  # type: ignore
+def _delayed_subsample_indices_block(
+    arr_chunk: NDArrayNum, subsample_indices: NDArrayNum, block_id: dict[str, Any]
+) -> NDArrayNum:
+    """Return 2D indices from the subsampled 1D valid indices per block."""
+
+    # Unravel indices of valid data to the shape of the block
+    ix, iy = np.unravel_index(np.argwhere(np.isfinite(arr_chunk.flatten()))[subsample_indices], shape=arr_chunk.shape)
+
+    # Convert to full-array indexes by adding the row and column starting indexes for this block
+    ix += block_id["xstart"]
+    iy += block_id["ystart"]
+
+    return np.hstack((ix, iy))
+
+
+def delayed_subsample(
```

Comment:
It would be great to compare at least runtime between the non delayed and delayed approaches, for arrays that fit in memory. To check that the method does not add too much overhead for that functionality.

Reply:
To check runtime we'd need to compare to `Raster.interp_points`, `Raster.reproject` etc, which currently live in the raster class directly and not as a separate function, so it's impractical.
This change in architecture needs to change for #383 anyway, opening two issues.